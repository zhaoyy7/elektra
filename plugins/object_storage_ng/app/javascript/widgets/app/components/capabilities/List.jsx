import React from "react"
import PropTypes from "prop-types"
import { createUseStyles } from "react-jss"

const useStyles = createUseStyles({
  missingCapability: {
    opacity: 0.25,
  },
})

const Capabilities = ({ data }) => {
  const classes = useStyles()
  const additionalCaps = React.useMemo(() => {
    const parsedCaps = [
      "id",
      "swift",
      "account_quotas",
      "bulk_delete",
      "bulk_upload",
      "container_quotas",
      "slo",
      "container_sync",
      "ratelimit",
      "staticweb",
      "tempurl",
    ]
    return Object.keys(data).filter((k) => parsedCaps.indexOf(k) < 0)
  }, [data])

  if (!data) return <span>No Capabilities available!</span>

  return (
    <div style={{ color: "#666" }}>
      <h5>Limits</h5>
      <ul className="fa-ul" id="limits">
        {data.swift && (
          <React.Fragment>
            <li>
              <span className="fa-li fa fa-check" />
              <strong>Max file size: </strong>
              {Math.floor(data.swift["max_file_size"] / 1024 / 1024 / 1024)} GB
            </li>
            <li>
              <span className="fa-li fa fa-check" />
              <strong>Max container name length: </strong>
              {data.swift["max_container_name_length"]}
            </li>
            <li>
              <span className="fa-li fa fa-check" />
              <strong>Max object name length: </strong>
              {data.swift["max_object_name_length"]}
            </li>
            <li>
              <span className="fa-li fa fa-check" />
              <strong>Container listing limit: </strong>
              {data.swift["container_listing_limit"]}
            </li>
          </React.Fragment>
        )}

        {data.bulk_upload && (
          <li>
            <span className="fa-li fa fa-check" />
            <strong>Max containers per extraction: </strong>
            {data.bulk_upload["max_containers_per_extraction"]}{" "}
            <small>(bulk upload)</small>
          </li>
        )}
        {data.bulk_delete && (
          <li>
            <span className="fa-li fa fa-check" />
            <strong>Max deletes per request: </strong>
            {data.bulk_delete["max_deletes_per_request"]}{" "}
            <small>(bulk delete)</small>
          </li>
        )}
        {data.slo && (
          <li>
            <span className="fa-li fa fa-check" />
            <strong>Max segments: </strong>
            {data.slo["max_manifest_segments"]}{" "}
            <small>(static large object)</small>
          </li>
        )}
      </ul>
      <h5>Capabilities</h5>
      <ul className="fa-ul" id="capabilities_list">
        <li className={data.account_quotas ? "" : classes.missingCapability}>
          <span className="fa-li fa fa-area-chart" />
          <strong>Account quotas </strong>
        </li>
        <li className={data.container_quotas ? "" : classes.missingCapability}>
          <span className="fa-li fa fa-area-chart" />
          <strong>Container quotas </strong>
        </li>
        <li className={data.ratelimit ? "" : classes.missingCapability}>
          <span className="fa-li fa fa-hand-stop-o" />
          API
          <strong> rate limiting </strong>
        </li>
        <li className={data.bulk_upload ? "" : classes.missingCapability}>
          <span className="fa-li fa fa-file-archive-o" />
          <strong>Bulk upload </strong>
          of archive files
        </li>
        <li className={data.bulk_delete ? "" : classes.missingCapability}>
          <span className="fa-li fa fa-eraser" />
          Efficient
          <strong> bulk deletion </strong>
        </li>
        <li className={data.container_sync ? "" : classes.missingCapability}>
          <span className="fa-li fa fa-refresh" />
          <strong>Container syncing </strong>
        </li>
        <li className={data.slo ? "" : classes.missingCapability}>
          <span className="fa-li fa fa-object-group" />
          Handling of large objects using
          <strong> static large object </strong>
          manifests
        </li>
        <li className={data.staticweb ? "" : classes.missingCapability}>
          <span className="fa-li fa fa-globe" />
          Distribution of container contents as
          <strong> static websites </strong>
        </li>
        <li className={data.tempurl ? "" : classes.missingCapability}>
          <span className="fa-li fa fa-clock-o" />
          Sharing of objects using
          <strong> temporary URLs </strong>
        </li>

        {additionalCaps.length > 0 && (
          <li>
            <span className="fa-li fa fa-ellipsis-v" />
            Additional capabilities:{" "}
            <strong>{additionalCaps.join(", ")}</strong>
          </li>
        )}
      </ul>

      <p className="help-block">
        For a more detailed report, use the
        <code>swift info</code>
        command on the
        <a href="#">Web Shell</a>.
      </p>
    </div>
  )
}

Capabilities.propTypes = {
  data: PropTypes.object.isRequired,
}

export default Capabilities
