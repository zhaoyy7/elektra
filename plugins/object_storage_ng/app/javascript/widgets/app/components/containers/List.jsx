import React from "react"
import PropTypes from "prop-types"
import { useGlobalState } from "../../StateProvider"
import useActions from "../../hooks/useActions"
import { Link, useHistory } from "react-router-dom"
import ItemsCount from "../shared/ItemsCount"
import TimeAgo from "../shared/TimeAgo"
import { Unit } from "lib/unit"
const unit = new Unit("B")
import { policy } from "lib/policy"
import { SearchField } from "lib/components/search_field"
import CapabilitiesPopover from "../capabilities/Popover"
import VirtualizedTable from "lib/components/VirtualizedTable"
import ContextMenu from "lib/components/ContextMenuPopover"
import Breadcrumb from "../shared/Breadcrumb"

const Table = ({ data, onMenuAction }) => {
  const columns = React.useMemo(
    () => [
      {
        label: "Container name",
        accessor: "name",
        sortable: "text",
        // filterable: true,
      },
      {
        label: "Last modified",
        accessor: "last_modified",
        width: "20%",
        sortable: "date",
      },
      {
        label: "Total size",
        accessor: "bytes",
        width: "20%",
        sortable: true,
      },
      {
        width: "60px",
      },
    ],
    []
  )

  const permissions = React.useMemo(() => {
    if (!policy) return {}
    return {
      canDelete: policy.isAllowed("object_storage_ng:container_delete"),
      canEmpty: policy.isAllowed("object_storage_ng:container_empty"),
      canShow: policy.isAllowed("object_storage_ng:container_get"),
      canShowAccessControl: policy.isAllowed(
        "object_storage_ng:container_show_access_control"
      ),
    }
  }, [policy])

  const Row = React.useCallback(({ Row, item }) => {
    return (
      <Row>
        <Row.Column>
          <span className="fa fa-fw fa-hdd-o" title="Container" />{" "}
          <Link to={`/containers/${item.name}/objects`} title="List Containers">
            {item.name}
          </Link>
          <br />
          <ItemsCount count={item.count} />
        </Row.Column>
        <Row.Column>
          <TimeAgo date={item.last_modified} originDate />
        </Row.Column>
        <Row.Column>{unit.format(item.bytes)}</Row.Column>
        <Row.Column>
          <ContextMenu>
            {permissions.canShow && (
              <ContextMenu.Item
                onClick={() => onMenuAction("properties", item)}
              >
                Properties
              </ContextMenu.Item>
            )}
            {permissions.canShowAccessControl && (
              <ContextMenu.Item
                onClick={() => onMenuAction("accessControl", item)}
              >
                Access Control
              </ContextMenu.Item>
            )}
            {(permissions.canShow || permissions.canShowAccessControl) && (
              <ContextMenu.Item divider />
            )}
            {permissions.canEmpty && (
              <ContextMenu.Item onClick={() => onMenuAction("empty", item)}>
                Empty
              </ContextMenu.Item>
            )}
            {permissions.canDelete && (
              <ContextMenu.Item onClick={() => onMenuAction("delete", item)}>
                Delete
              </ContextMenu.Item>
            )}
          </ContextMenu>
        </Row.Column>
      </Row>
    )
  }, [])

  return (
    <VirtualizedTable
      height="max"
      rowHeight={50}
      columns={columns}
      data={data || []}
      renderRow={Row}
      showHeader
      bottomOffset={160} // footer height
    />
  )
}

Table.propTypes = {
  data: PropTypes.arrayOf(PropTypes.object),
  onMenuAction: PropTypes.func,
}

const List = () => {
  const containers = useGlobalState("containers")
  const { loadContainersOnce } = useActions()
  const history = useHistory()
  const [searchTerm, updateSearchTerm] = React.useState("")

  React.useEffect(() => {
    loadContainersOnce()
  }, [loadContainersOnce])

  const handleMenuAction = React.useCallback(
    (action, item) => {
      switch (action) {
        case "accessControl":
          return history.push(`/containers/${item.name}/access-control`)
        case "empty":
          return history.push(`/containers/${item.name}/empty`)
        case "delete":
          return history.push(`/containers/${item.name}/delete`)
        case "properties":
          return history.push(`/containers/${item.name}/properties`)
      }
    },
    [history.push]
  )

  const items = React.useMemo(
    () =>
      containers.items.filter(
        (i) => !searchTerm || (i.name && i.name.indexOf(searchTerm) >= 0)
      ),
    [searchTerm, containers.items]
  )

  return (
    <React.Fragment>
      <div className="toolbar">
        <SearchField
          onChange={(term) => updateSearchTerm(term)}
          placeholder="name"
          text="Filters by name"
        />

        <div className="main-buttons">
          {policy.isAllowed("object_storage_ng:container_create") && (
            <React.Fragment>
              <CapabilitiesPopover />
              <button
                className="btn btn-link"
                onClick={() => loadContainersOnce({ reload: true })}
              >
                <i className="fa fa-refresh" />
              </button>
              <Link to="/containers/new" className="btn btn-primary">
                Create container
              </Link>
            </React.Fragment>
          )}
        </div>
      </div>
      {!policy.isAllowed("object_storage_ng:container_list") ? (
        <span>You are not allowed to see this page</span>
      ) : containers.isFetching ? (
        <span>
          <span className="spinner" />
          Loading...
        </span>
      ) : items.length === 0 ? (
        <span>No Containers found.</span>
      ) : (
        <div>
          <Breadcrumb count={items.length} />
          <Table data={items} onMenuAction={handleMenuAction} />
        </div>
      )}
    </React.Fragment>
  )
}

export default List
