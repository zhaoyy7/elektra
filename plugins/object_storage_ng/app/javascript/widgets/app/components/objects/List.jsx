import React from "react"
import PropTypes from "prop-types"
import {
  useParams,
  Link,
  useRouteMatch,
  useHistory,
  Route,
} from "react-router-dom"
import Breadcrumb from "../shared/Breadcrumb"
import useUrlParamEncoder from "../../hooks/useUrlParamEncoder"
import { Alert } from "react-bootstrap"
import { SearchField } from "lib/components/search_field"
import useActions from "../../hooks/useActions"
import NewObject from "./New"
import UploadFile from "./Upload"
import ShowProperties from "./Show"
import CopyFile from "./Copy"
import DownloadInstructions from "./DownloadInstructions"

import { reducer, initialState } from "./reducer"
import Table from "./Table"
import { LIMIT } from "./config"
import { confirm } from "lib/dialogs"

const Objects = ({ objectStoreEndpoint }) => {
  let { url } = useRouteMatch()
  let objectsRoot = url.replace(/([^/])\/objects.*/, "$1/objects")
  let history = useHistory()
  let { name: containerName, objectPath } = useParams()
  const {
    value: currentPath,
    encode,
    getFileName,
  } = useUrlParamEncoder(objectPath)
  const [searchTerm, setSearchTerm] = React.useState(null)

  const {
    loadContainerObjects,
    loadObjectMetadata,
    deleteObject,
    deleteObjects,
    loadAccountMetadataOnce,
    downloadObject,
    rawObjectUrl,
  } = useActions()

  const [objects, dispatch] = React.useReducer(reducer, initialState)

  const loadObjects = React.useCallback(() => {
    // Load objects
    // For objects beginning with a slash, this function is called recursively
    // until there are no objects beginning with a slash

    /** Normally, slashes are intended as delimiters for the directories. 
     * However, it is also allowed to create objects with leading slashes in names. 
     * This leads to a problem when we call the API with prefix and delimiter.
     * Example:
     * Given: [
     *  {name: "test/sub_test/image.pmg"},
     *  {name: "/test1/image.png"},
     *  {name: "//test3/sub_test3/a/b.png"}
     * ]
     * API Call: prefix: "", delimiter: "/" => ["test/", "/", "//"]
     * API Call: prefix: "/", delimiter: "/" => ["/test1/", "//"]
     * API Call: prefix: "//", delimiter: "/" => ["//test3/"]

    * As you can see, all calls deliver different results. To get all objects, 
    * even those starting with multiple slashes, we start with the empty prefix and 
    * * load the objects. After that, we search the results for names that only contain slashes. 
    * Remove these and recursively load with the prefix of the removed items, etc. 
    * until the results contain no objects with leading slashes.
    */
    const loadAllObjects = async (prefix = "") => {
      let objects = await loadContainerObjects(containerName, {
        prefix,
        delimiter: "/",
      }).then(({ data }) => data)
      // find index of the first object which name starts with a slash
      let regex = new RegExp(`^${prefix}/+$`)
      const startingWithSlashIndex = objects.findIndex(
        (o) => o.subdir && o.subdir.match(regex)
      )

      // index not found -> end of recursion
      if (startingWithSlashIndex < 0)
        return objects.filter((o) => o.name !== prefix)

      // get the new prefix based on the found object
      const newPrefix = objects[startingWithSlashIndex].subdir
      // remove all objects which names start with multiple slashes
      objects = objects.filter((o) => !(o.name || o.subdir).match(regex))
      // load objects recursively based on the new prefix
      let objectsStartingWithSlash = await loadAllObjects(newPrefix)
      // add new objects to the root objects
      let newObjects = objects.concat(objectsStartingWithSlash)

      // remove duplicates
      return newObjects.filter(
        (item, index) => newObjects.indexOf(item) === index
      )
    }

    dispatch({ type: "REQUEST_ITEMS" })
    loadAllObjects(currentPath)
      .then((items) => {
        // extend items with display_name and sort
        items.forEach((i) => {
          // display name
          let dn = (i.name || i.subdir).replace(currentPath, "")
          if (dn[dn.length - 1] === "/") dn = dn.slice(0, -1)
          i.display_name = dn
        })
        items = items.sort((a, b) =>
          a.display_name > b.display_name
            ? 1
            : a.display_name < b.display_name
            ? -1
            : 0
        )
        dispatch({ type: "RECEIVE_ITEMS", items })
      })
      .catch((error) =>
        dispatch({ type: "RECEIVE_ERROR", error: error.message })
      )
  }, [containerName, currentPath, dispatch, loadContainerObjects])

  React.useEffect(() => {
    loadObjects()
  }, [loadObjects])

  // Return the delete function and a cancel function
  const [deleteFolder, cancelDeleteFolder] = React.useMemo(() => {
    // this variable indicates whether deletion is active
    let active

    // this is the function which deletes all objects inside a folder
    const action = (name, options = {}) => {
      confirm("This folder will be irrevocably deleted", {
        confirmLabel: "Confirm",
        abortLabel: "Cancel",
      })
        .then(() => {
          const currentContainerName = options.containerName || containerName
          if (!currentContainerName || !name) return
          active = true

          // This function deletes all objects of a folder.
          // Since the number of objects to be loaded and deleted is limited,
          // we delete the objects in chunks.
          const deleteAllObjects = async () => {
            let marker
            let deletedCount = 0
            let processing = true
            // We load objects, delete them and repeat this process until there are no more objects
            while (active && processing) {
              // use prefix to limit the deletion to current folder
              await loadContainerObjects(currentContainerName, {
                marker,
                prefix: name,
              }).then(async ({ data }) => {
                if (data.length > 0 && active) {
                  // delete objects
                  await deleteObjects(currentContainerName, data)
                  // update progress
                  dispatch({
                    type: "UPDATE_ITEM",
                    name,
                    progress: (deletedCount += data.length),
                  })
                  // marker is the last item. Marker is used to load the next chunk of items.
                  marker = data.pop().name
                } else {
                  processing = false
                }
              })
            }
          }

          dispatch({ type: "UPDATE_ITEM", name, isDeleting: true })
          deleteAllObjects()
            .then(() => dispatch({ type: "REMOVE_ITEM", name }))
            .catch((error) => {
              if (!active) return
              dispatch({
                type: "UPDATE_ITEM",
                name,
                isDeleting: false,
                error: error.message,
              })
            })
        })
        .catch(() => null)
    }

    // return the actual action and a cancel function to cancel the delete process for large containers
    return [action, () => (active = false)]
  }, [
    containerName,
    loadContainerObjects,
    deleteObjects,
    deleteObjects,
    dispatch,
  ])

  // Delete a single file
  const deleteFile = React.useCallback(
    (name, options = {}) => {
      confirm("This object will be irrevocably deleted", {
        confirmLabel: "Confirm",
        abortLabel: "Cancel",
      })
        .then(() => {
          const deleteFunc = async () => {
            if (options.keepSegments !== false) {
              return deleteObject(containerName, name)
            } else {
              const metadata = await loadObjectMetadata(containerName, name)
              if (metadata["x-static-large-object"]) {
                // SLO
                return deleteObject(containerName, name, {
                  "multipart-manifest": "delete",
                })
              } else if (metadata["x-object-manifest"]) {
                // DLO
                // delete dlo manifest
                const [segmentContainer, segmentObject] =
                  metadata["x-object-manifest"].split("/")
                return deleteObject(containerName, name).then(() =>
                  deleteFolder(segmentObject, {
                    containerName: segmentContainer,
                  })
                )
              } else {
                return deleteObject(containerName, name)
              }
            }
          }

          dispatch({ type: "UPDATE_ITEM", name, isDeleting: true })
          //deleteObject(containerName, name)
          deleteFunc()
            .then(() => dispatch({ type: "REMOVE_ITEM", name }))
            .catch((error) =>
              dispatch({
                type: "UPDATE_ITEM",
                name,
                isDeleting: false,
                error: error.message,
              })
            )
        })
        .catch(() => null)
    },
    [containerName, dispatch, deleteObject, deleteFolder, loadObjectMetadata]
  )

  const downloadFile = React.useCallback(
    (name, size = 0) => {
      if (size < LIMIT) {
        // less than LIMIT
        let fileName = getFileName(name)
        dispatch({ type: "UPDATE_ITEM", name, isProcessing: true })
        downloadObject(containerName, name, { fileName })
          .then(() =>
            dispatch({ type: "UPDATE_ITEM", name, isProcessing: false })
          )
          .catch((error) =>
            dispatch({
              type: "UPDATE_ITEM",
              name,
              isProcessing: false,
              error: error.message,
            })
          )
      } else {
        history.replace(
          `${url}/${objectPath ? "" : encode("") + "/"}${encodeURIComponent(
            name
          )}/download-instructions`
        )
      }
    },
    [
      containerName,
      objectPath,
      loadAccountMetadataOnce,
      downloadObject,
      getFileName,
      dispatch,
      history,
    ]
  )

  // cancel current deletion process
  React.useEffect(() => {
    return () => cancelDeleteFolder && cancelDeleteFolder()
  }, [cancelDeleteFolder])

  const changeDir = React.useCallback(
    (subdir) => history.push(`${objectsRoot}/${encode(subdir)}`),
    [history, objectsRoot]
  )

  const showProperties = React.useCallback(
    (name) => {
      history.push(
        `${url}/${objectPath ? "" : encode("") + "/"}${encodeURIComponent(
          name
        )}/show`
      )
    },
    [history, objectPath, url]
  )

  const moveFile = React.useCallback(
    (name) => {
      history.push(
        `${url}/${objectPath ? "" : encode("") + "/"}${encodeURIComponent(
          name
        )}/move`
      )
    },
    [history, objectPath, url]
  )

  const copyFile = React.useCallback(
    (name) => {
      history.push(
        `${url}/${objectPath ? "" : encode("") + "/"}${encodeURIComponent(
          name
        )}/copy`
      )
    },
    [history, objectPath, url]
  )

  const handleFolderCreated = React.useCallback(
    (values) => {
      dispatch({ type: "RECEIVE_ITEM", item: values })
    },
    [dispatch]
  )

  const filteredItems = React.useMemo(() => {
    let items = objects.items

    if (searchTerm && searchTerm.length > 0) {
      items = objects.items.filter(
        (i) => i.display_name.indexOf(searchTerm) >= 0
      )
    }
    return items.sort((a, b) => {
      if (a.subdir && !b.subdir) return -1
      if (!a.subdir && b.subdir) return 1
      if (a.display_name > b.display_name) return 1
      if (a.display_name < b.display_name) return -1
      return 0
    })
  }, [objects.items, searchTerm])

  return (
    <React.Fragment>
      <Route exact path="/containers/:name/objects/:objectPath?/new">
        <NewObject onCreated={handleFolderCreated} />
      </Route>
      <Route exact path="/containers/:name/objects/:objectPath?/upload">
        <UploadFile
          refresh={loadObjects}
          objectStoreEndpoint={objectStoreEndpoint}
        />
      </Route>
      <Route exact path="/containers/:name/objects/:objectPath?/:object/move">
        <CopyFile refresh={loadObjects} deleteAfter />
      </Route>
      <Route exact path="/containers/:name/objects/:objectPath?/:object/copy">
        <CopyFile refresh={loadObjects} showCopyMetadata />
      </Route>
      <Route exact path="/containers/:name/objects/:objectPath?/:object/show">
        <ShowProperties objectStoreEndpoint={objectStoreEndpoint} />
      </Route>
      <Route
        exact
        path="/containers/:name/objects/:objectPath?/:object/download-instructions"
      >
        <DownloadInstructions objectStoreEndpoint={objectStoreEndpoint} />
      </Route>

      <div className="toolbar">
        <SearchField
          onChange={setSearchTerm}
          placeholder="name"
          text="Filters by name"
        />

        <div className="main-buttons">
          <Link className="btn btn-default" to={`${url}/new`}>
            Create folder
          </Link>
          <Link className="btn btn-primary" to={`${url}/upload`}>
            Upload file
          </Link>
        </div>
      </div>
      <Breadcrumb count={filteredItems.length} />
      {objects.isFetching ? (
        <span>
          <span className="spinner" /> Loading...
        </span>
      ) : objects.error ? (
        <Alert bsStyle="danger">{objects.error}</Alert>
      ) : !objects || objects.items.length === 0 ? (
        <span>No entries found.</span>
      ) : (
        <>
          {filteredItems.length > 0 ? (
            <Table
              data={filteredItems}
              changeDir={(item) => changeDir(item.subdir)}
              deleteFile={(item) => deleteFile(item.name)}
              deleteFolder={(item) => deleteFolder(item.subdir)}
              downloadFile={(item) => downloadFile(item.name, item.bytes)}
              showProperties={(item) => showProperties(item.name)}
              rawUrl={(item) =>
                rawObjectUrl(containerName, item.name, {
                  inline: 1,
                })
              }
              copyFile={(item) => copyFile(item.name)}
              moveFile={(item) => moveFile(item.name)}
            />
          ) : (
            <span>This folder is empty</span>
          )}
        </>
      )}
    </React.Fragment>
  )
}

Objects.propTypes = {
  objectStoreEndpoint: PropTypes.string,
}

export default Objects
