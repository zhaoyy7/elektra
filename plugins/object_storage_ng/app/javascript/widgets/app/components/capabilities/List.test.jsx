import React from "react"
import renderer from "react-test-renderer"
import List from "./List"

describe("router", () => {
  test("routes", () => {
    const component = renderer.create(<List data={{}} />)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
