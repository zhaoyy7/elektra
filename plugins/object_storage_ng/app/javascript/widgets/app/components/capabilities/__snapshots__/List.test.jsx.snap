// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`router routes 1`] = `
<div
  style={
    Object {
      "color": "#666",
    }
  }
>
  <h5>
    Limits
  </h5>
  <ul
    className="fa-ul"
    id="limits"
  />
  <h5>
    Capabilities
  </h5>
  <ul
    className="fa-ul"
    id="capabilities_list"
  >
    <li
      className="missingCapability-0-2-1"
    >
      <span
        className="fa-li fa fa-area-chart"
      />
      <strong>
        Account quotas 
      </strong>
    </li>
    <li
      className="missingCapability-0-2-1"
    >
      <span
        className="fa-li fa fa-area-chart"
      />
      <strong>
        Container quotas 
      </strong>
    </li>
    <li
      className="missingCapability-0-2-1"
    >
      <span
        className="fa-li fa fa-hand-stop-o"
      />
      API
      <strong>
         rate limiting 
      </strong>
    </li>
    <li
      className="missingCapability-0-2-1"
    >
      <span
        className="fa-li fa fa-file-archive-o"
      />
      <strong>
        Bulk upload 
      </strong>
      of archive files
    </li>
    <li
      className="missingCapability-0-2-1"
    >
      <span
        className="fa-li fa fa-eraser"
      />
      Efficient
      <strong>
         bulk deletion 
      </strong>
    </li>
    <li
      className="missingCapability-0-2-1"
    >
      <span
        className="fa-li fa fa-refresh"
      />
      <strong>
        Container syncing 
      </strong>
    </li>
    <li
      className="missingCapability-0-2-1"
    >
      <span
        className="fa-li fa fa-object-group"
      />
      Handling of large objects using
      <strong>
         static large object 
      </strong>
      manifests
    </li>
    <li
      className="missingCapability-0-2-1"
    >
      <span
        className="fa-li fa fa-globe"
      />
      Distribution of container contents as
      <strong>
         static websites 
      </strong>
    </li>
    <li
      className="missingCapability-0-2-1"
    >
      <span
        className="fa-li fa fa-clock-o"
      />
      Sharing of objects using
      <strong>
         temporary URLs 
      </strong>
    </li>
  </ul>
  <p
    className="help-block"
  >
    For a more detailed report, use the
    <code>
      swift info
    </code>
    command on the
    <a
      href="#"
    >
      Web Shell
    </a>
    .
  </p>
</div>
`;
